pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -0,0 +1,347 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package memorymanager
+
+import (
+	"fmt"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog"
+	corehelper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
+	"k8s.io/kubernetes/pkg/kubelet/cm/memorymanager/state"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
+)
+
+const policyTypeSingleNUMA policyType = "single-numa"
+
+// TODO: need to implement all methods
+
+type ReservedMemory map[int]map[v1.ResourceName]uint64
+
+// SingleNUMAPolicy is implementation of the policy interface for the single NUMA policy
+// TODO: need to re-evaluate what field we really need
+type singleNUMAPolicy struct {
+	// machineInfo contains machine memory related information
+	machineInfo *cadvisorapi.MachineInfo
+	// reserved contains memory that reserved for kube
+	systemReserved ReservedMemory
+	// topology manager reference to get container Topology affinity
+	affinity topologymanager.Store
+}
+
+var _ Policy = &singleNUMAPolicy{}
+
+// NewPolicySingleNUMA returns new single NUMA policy instance
+func NewPolicySingleNUMA(machineInfo *cadvisorapi.MachineInfo, reserved ReservedMemory, affinity topologymanager.Store) Policy {
+	// TODO: check if we have enough reserved memory for the system
+	//for _, node := range reserved {
+	//	if node[v1.ResourceMemory] == 0 {
+	//
+	//	}
+	//}
+	return &singleNUMAPolicy{
+		machineInfo:    machineInfo,
+		systemReserved: reserved,
+		affinity:       affinity,
+	}
+}
+
+func (p *singleNUMAPolicy) Name() string {
+	return string(policyTypeSingleNUMA)
+}
+
+func (p *singleNUMAPolicy) Start(s state.State) error {
+	if err := p.validateState(s); err != nil {
+		klog.Errorf("[memorymanager] Invalid state: %v, please drain node and remove policy state file", err)
+		return err
+	}
+	return nil
+}
+
+// Allocate call is idempotent
+func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) error {
+	// allocate the memory only for guaranteed pods
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	klog.Infof("[memorymanager] Allocate (pod: %s, container: %s)", pod.Name, container.Name)
+	if blocks := s.GetMemoryBlocks(string(pod.UID), container.Name); blocks != nil {
+		klog.Infof("[memorymanager] Container already present in state, skipping (pod: %s, container: %s)", pod.Name, container.Name)
+		return nil
+	}
+
+	// Call Topology Manager to get the aligned memory affinity across all hint providers.
```

Comment:
I think we get aligned affinity for all requested resource here.
// Call Topology Manager to get the aligned memory affinity across all hint providers.
=>
// Call Topology Manager to get the aligned affinity across all hint providers.
// Call Topology Manager to get the aligned compute resource affinity across all hint providers.